### PR TITLE
Gcp cloud functions

### PR DESCRIPTION
**Pull request by CourbeB**

### CODE REVIEW

File: technologies/job/gcp-cloud-functions/project/instanceActions.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
+const { Response, JobStatus, Log } = require('@saagie/sdk');
+const { google } = require('googleapis');
+const cloudfunctions = google.cloudfunctions('v1');
+const logging = google.logging('v2');
+const { getConnexion } = require('./utils');
```
Comment (yoannfleurydev) [inline technologies/job/gcp-cloud-functions/project/instanceActions.js:5]:
```suggestion
const { getAuth } = require('./utils');
```

File: technologies/job/gcp-cloud-functions/project/instanceActions.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  try {
+    const gcpKey = JSON.parse(job.featuresValues.endpoint.jsonKey);
+
+    authClient = getConnexion(gcpKey);
+
+    console.log(`Retrieve status for ${job.featuresValues.function.id}`)
+    const { data } = await cloudfunctions.projects.locations.functions.get({
+      auth: authClient,
+      name : `${job.featuresValues.function.id}`,
+    });
```
Comment (yoannfleurydev) [inline technologies/job/gcp-cloud-functions/project/instanceActions.js:39]:
```suggestion
    const auth = getAuth(gcpKey);

    console.log(`Retrieve status for ${job.featuresValues.function.id}`)
    const { data } = await cloudfunctions.projects.locations.functions.get({
      auth,
      name : `${job.featuresValues.function.id}`,
    });
```

File: technologies/job/gcp-cloud-functions/project/instanceActions.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    switch (data.status) {
+      case 'ACTIVE':
+        return Response.success(JobStatus.RUNNING);
+      case 'OFFLINE':
+        return Response.success(JobStatus.FAILED);
+      case 'DEPLOY_IN_PROGRESS':
+        return Response.success(JobStatus.AWAITING);
+      default:
+        return Response.success(JobStatus.KILLED);
+    }
```
Comment (yoannfleurydev) [inline technologies/job/gcp-cloud-functions/project/instanceActions.js:50]:
```suggestion
    const JOB_STATUS = {
        'ACTIVE': JobStatus.RUNNING,
        'OFFLINE': JobStatus.FAILED,
        'DEPLOY_IN_PROGRESS': JobStatus.AWAITING,
    };

    return Response.success(JOB_STATUS[data.status] || JobStatus.KILLED);
```

Useful to directly see which Saagie status correspond to which Google Cloud Functions status.
Using `||`, if `JOB_STATUS[data.status]` does not exists, then `JobStatus.KILLED` will be the default.

File: technologies/job/gcp-cloud-functions/project/jobForm.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
+const { Response } = require('@saagie/sdk');
+const { google } = require('googleapis');
+const cloudfunctions = google.cloudfunctions('v1');
+const { getConnexion } = require('./utils');
```
Comment (yoannfleurydev) [inline technologies/job/gcp-cloud-functions/project/jobForm.js:4]:
```suggestion
const { getAuth } = require('./utils');
```

File: technologies/job/gcp-cloud-functions/project/jobForm.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+ */
+exports.getFunctions = async ({ featuresValues }) => {
+  const gcpKey = JSON.parse(featuresValues.endpoint.jsonKey);
+
+  try {
+    authClient = getConnexion(gcpKey);
+    const { data: { functions } } = await cloudfunctions.projects.locations.functions.list({
+      auth: authClient,
+      parent : `projects/saagie-internal-testing/locations/${featuresValues.endpoint.location.id}`,
+    });
```
Comment (yoannfleurydev) [inline technologies/job/gcp-cloud-functions/project/jobForm.js:19]:
```suggestion
    const auth = getAuth(gcpKey);
    const { data: { functions } } = await cloudfunctions.projects.locations.functions.list({
      auth,
      parent : `projects/saagie-internal-testing/locations/${featuresValues.endpoint.location.id}`,
    });
```

File: technologies/job/gcp-cloud-functions/project/instanceActions.js [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+    authClient = getConnexion(gcpKey);
+
+    const { data: { entries }} = await logging.entries.list({
+      requestBody: {
+        filter: `resource.type=cloud_function resource.labels.function_name=${job.featuresValues.function.label} resource.labels.region=${job.featuresValues.endpoint.location.id} log_name=projects/saagie-internal-testing/logs/cloudfunctions.googleapis.com%2Fcloud-functions`,
+        orderBy: "timestamp desc",
+        resourceNames: ["projects/saagie-internal-testing"]
+      }, 
+      auth: authClient
```
Comment (yoannfleurydev) [inline technologies/job/gcp-cloud-functions/project/instanceActions.js:74]:
```suggestion
    const auth = getAuth(gcpKey);

    const { data: { entries }} = await logging.entries.list({
      requestBody: {
        filter: `resource.type=cloud_function resource.labels.function_name=${job.featuresValues.function.label} resource.labels.region=${job.featuresValues.endpoint.location.id} log_name=projects/saagie-internal-testing/logs/cloudfunctions.googleapis.com%2Fcloud-functions`,
        orderBy: "timestamp desc",
        resourceNames: ["projects/saagie-internal-testing"]
      }, 
      auth
```

File: technologies/job/gcp-cloud-functions/project/context.yaml [hunk 7]
```diff
@@ -0,0 +1,47 @@
+id: project
+label: project
```
Comment (youenchene) [inline technologies/job/gcp-cloud-functions/project/context.yaml:2]: project is not the good name here, nothing nothing bound to gcp function UI.

default or functions would be better.

File: technologies/job/gcp-cloud-functions/project/context.yaml [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+      helper: "Full JSON document"
+    - type: SELECT
+      name: location
+      label: Location
+      required: true
+      options:
+        script: ./endpointForm.js
+        function: getLocations
+      dependsOn:
+        - jsonKey
```
Comment (youenchene) [inline technologies/job/gcp-cloud-functions/project/context.yaml:21]: This filter should be in the job form part.

May be optional for not so much functions contexts.

Comment (youenchene) [reply technologies/job/gcp-cloud-functions/project/context.yaml:21]: + we suggest "Region" instead of location to respect Google wording.

**Pull request merged**